contracts/PushStaking/PushStaking.sol

```diff
@@ -0,0 +1,549 @@
+        emit NewSharesIssued(_walletAddress, sharesToBeAllocated);
+    }
+    /*
+     * Removing Wallet Share from a Wallet
+     * 1. removes the shares from a wallet completely
+     * 2. Can be called by governance.
+     * 3. Updates WALLET_TOTAL_SHARES
+     * 4. Emits out an event.
+    */
+
+    function removeWalletShare(address _walletAddress) public onlyGovernance {
+        if(_walletAddress == address(0) || _walletAddress == FOUNDATION) {
+            revert Errors.InvalidArgument_WrongAddress(_walletAddress);
+        }
+        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
+        _adjustWalletAndTotalStake(_walletAddress, 0, sharesToBeRemoved);
+        _adjustWalletAndTotalStake(FOUNDATION, sharesToBeRemoved, 0);
+
+        emit SharesRemoved(_walletAddress, sharesToBeRemoved);
+    }
+
+    function decreaseWalletShare(
+        address _walletAddress,
+        GenericTypes.Percentage memory _percentage
+    )
+        external
+        onlyGovernance
+    {
+        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
+        removeWalletShare(_walletAddress);
+        addWalletShare(_walletAddress, _percentage);
+        emit SharesDecreased(_walletAddress, sharesToBeRemoved, walletShareInfo[_walletAddress].walletShare);
+    }
+
+    /*
+     *Reward Calculation for a Given Wallet
+     * 1. calculateWalletRewards(address wallet)
+     * 2. public helper function
+     * 3. Helps in calculating rewards for a specific wallet based on
+       * a. Their Wallet Share 
+        * b. Total Wallet Shares
+        * c. Total Rewards available in WALLET_TOTAL_SHARES
+     * 4. Once calculated rewards for a sepcific wallet can be updated in WalletToRewards mapping
+     * 5. Reward can be calculated for wallets similar they are calculated for Token holders in a specific epoch.
+        * Reward for a Given Wallet X = ( Wallet Share of X / WALLET_TOTAL_SHARES) * WALLET_FEE_POOL
+    */
+    //TODO logic yet to be finalized
+    function calculateWalletRewards(address _wallet, uint256 _epochId) public view returns (uint256) {
+        return (walletShareInfo[_wallet].epochToWalletShares[_epochId] * epochRewardsForWallets[_epochId])
+            / epochToTotalShares[_epochId];
+    }
+
+    function claimShareRewards() external returns (uint256 rewards) {
+        _adjustWalletAndTotalStake(msg.sender, 0, 0);
+
+        uint256 currentEpoch = lastEpochRelative(genesisEpoch, block.number);
+        uint256 nextFromEpoch = lastEpochRelative(genesisEpoch, walletShareInfo[msg.sender].lastClaimedBlock);
+
+        uint256 _tillEpoch = currentEpoch - 1;
+
+        for (uint256 i = nextFromEpoch; i <= _tillEpoch; i++) {
+            uint256 claimableReward = calculateWalletRewards(msg.sender, i);
+            rewards = rewards + claimableReward;
+        }
+
+        usersRewardsClaimed[msg.sender] = usersRewardsClaimed[msg.sender] + rewards;
```

Comment:
Medium Issue:

- The contract uses the same mapping ( usersRewardsClaimed ) for updating the claimed rewards for both Token Holder and Wallet Holder.
- It is possible that 0xABC is a Wallet Address assigned by the governance to addShares() but then 0xABC also becomes a staker as it holds the token.
- In such a case, the usersRewardsClaimed mapping holds rewards from two different incentive-sources and different reward criteras, which isn't ideal
- Note: The only reason this is _medium_ or _low_ for now is coz the usersRewardsClaimed is not used for any other logic other than just read-only purpose. If in future upgrades, this map plays a role in some logic or validation, it becomes _critical_.